lib/breadcrumbs.ts

```diff
@@ -0,0 +1,182 @@
+  Error.prepareStackTrace = originalPrepareStacktrace;
+  return callsite;
+}
+
+export function addBreadcrumb(
+  breadcrumb: string | Breadcrumb,
+  type: InternalBreadcrumb["type"] = "manual",
+) {
+  const crumbs = getBreadcrumbs();
+
+  debug("using breadcrum store:", crumbs);
+
+  if (!crumbs) {
+    return;
+  }
+
+  if (typeof breadcrumb === "string") {
+    const expandedBreadcrumb: Breadcrumb = {
+      message: breadcrumb,
+      level: "info",
+      category: "",
+    };
+
+    breadcrumb = expandedBreadcrumb;
+  }
+
+  const callsite = getCallsite();
+
+  const internalCrumb: InternalBreadcrumb = {
+    ...(breadcrumb as Breadcrumb),
+    category: breadcrumb.category || "",
+    message: breadcrumb.message || "",
+    level: breadcrumb.level || "info",
+    timestamp: Number(new Date()),
+    type,
+    className: callsite?.fileName,
+    methodName: callsite?.functionName,
+
+    // TODO - do we need to do any source mapping?
+    lineNumber: callsite?.lineNumber || undefined,
+  };
+
+  debug("[breadcrumbs.ts] recorded breadcrumb:", internalCrumb);
+
+  crumbs.push(internalCrumb);
+}
+
+export function addRequestBreadcrumb(request: Request, response: Response) {
+  debug(`Add request breadcrumb: ${request}`);
+  const crumbs = getBreadcrumbs();
+
+  if (!crumbs) {
+    debug("Add request breadcrumb skip, no store!");
+    return;
+  }
+
+  const internalCrumb: InternalBreadcrumb = {
+    category: "http",
+    message: `${request.method} ${request.url}`,
+    level: "info",
+    timestamp: Number(new Date()),
+    type: "request",
+  };
+
+  crumbs.push(internalCrumb);
+}
+
+export function getBreadcrumbs(): InternalBreadcrumb[] | null {
+  if (!asyncLocalStorage) {
+    return null;
+  }
+
+  const store = asyncLocalStorage.getStore();
+
+  if (store) {
+    return store;
+  }
+
+  const newStore: InternalBreadcrumb[] = [];
+
+  debug("enter with new store");
+  asyncLocalStorage.enterWith(newStore);
+
+  return newStore;
+}
+
+export function runWithBreadcrumbs(
+  f: () => void,
+  store: InternalBreadcrumb[] = [],
+) {
+  if (!asyncLocalStorage) {
+    f();
+    return;
+  }
+  debug("running function with breadcrumbs");
+
+  asyncLocalStorage.run(store, f);
+}
+
+export function clear() {
+  if (!asyncLocalStorage) {
+    return;
+  }
+
+  debug("clearing stored breadcrumbs, entering with new store");
+  asyncLocalStorage.enterWith([]);
+}
+
+// const consoleMethods: [keyof typeof console, InternalBreadcrumb["level"]][] = [
+//   ["debug", "debug"],
+//   ["log", "info"],
+//   ["info", "info"],
+//   ["warn", "warning"],
+//   ["error", "error"],
+// ];
+
+// for (const [method, level] of consoleMethods) {
+//   const oldMethod = (console as any)[method];
+//   (console as any)[method] = function logWithBreadcrumb<T>(
+//     this: T,
+//     ...args: any[]
+//   ) {
+//     addBreadcrumb({ message: args.join(" "), level }, "console");
+//     return oldMethod.apply(this, args);
+//   };
+// }
```

Comment:
Old code from the original `breadcrumb.ts`. It overrites the `console.` methods so they log automatically into breadcrumbs. I'll probably end up removing it, users should be able to do that themselves.

Reply:
I think at first glance this looks like a nice feature that could be offered via a setting/config.

Reply:
To move things along, maybe create a spin-off ticket to track my "feature suggestion" and then remove the commented code from here for this PR's purpose.

Reply:
Done: https://github.com/MindscapeHQ/raygun4node/issues/216